netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -258,51 +297,132 @@ public void operationComplete(ChannelFuture future) throws Exception {
         ch.closeFuture().addListener(loopReleaser);
       }
     });
-    // Bind and start to accept incoming connections.
-    ChannelFuture future = b.bind(address);
-    // We'd love to observe interruption, but if interrupted we will need to close the channel,
-    // which itself would need an await() to guarantee the port is not used when the method returns.
-    // See #6850
-    future.awaitUninterruptibly();
-    if (!future.isSuccess()) {
-      throw new IOException(String.format("Failed to bind to address %s", address), future.cause());
+
+    Future<Map<ChannelFuture, SocketAddress>> bindCallFuture =
+        bossExecutor.submit(new Callable<Map<ChannelFuture, SocketAddress>>() {
+          @Override
+          public Map<ChannelFuture, SocketAddress> call() {
+            Map<ChannelFuture, SocketAddress> map = new HashMap<>();
+            for (SocketAddress address: addresses) {
+                ChannelFuture future = b.bind(address);
+                map.put(future, address);
+            }
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures =
+        bindCallFuture.awaitUninterruptibly().getNow();
+    if (channelFutures == null) {
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), bindCallFuture.cause());
     }
-    channel = future.channel();
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+    BindResult bindResult = new BindResult();
+    for (Map.Entry<ChannelFuture, SocketAddress> entry: channelFutures.entrySet()) {
+      // We'd love to observe interruption, but if interrupted we will need to close the channel,
+      // which itself would need an await() to guarantee the port is not used when the method
+      // returns. See #6850
+      ChannelFuture future = entry.getKey().awaitUninterruptibly();
+      if (!future.isSuccess()) {
+        bindResult.failAddress(entry.getValue()).causedBy(future.cause());
+      } else {
+        final Channel channel = future.channel();
+        channel.eventLoop().execute(new Runnable() {
+          @Override
+          public void run() {
+            InternalInstrumented<SocketStats> listenSocketStats = new ListenSocket(channel);
+            channelz.addListenSocket(listenSocketStats);
+            listenSocketStatsList.add(listenSocketStats);
+          }
+        });
+        channelGroup.add(channel);
       }
-    });
+    }
+
+    if (!bindResult.isSuccess()) {
+      if (!channelGroup.isEmpty()) {
+        closeChannelGroup().awaitUninterruptibly();
```

Comment:
I see that closeChannelGroup() was split out to be reused here, but it doesn't seem safe to call. It does inappropriate things in this context like sharedResourceReferenceCounter.release() and protocolNegotiator.close(). (There's pre-existing issues already surrounding sharedResourceReferenceCounter, but this change doesn't fix them so it seems better to leave it alone.)

If listenSocketStatsList is removed by my other suggestion, I think this just becomes `channelGroup.close().awaitUninterruptibly()`.